### PR TITLE
Feat: Eagle3 HF Online - support nemotron models

### PR DESCRIPTION
**Pull request by h-guo18**
## What does this PR do?

**Type of change:** New feature 

**Overview:** 
- Support the nano and nano-VL in eagle3 online mode:
  - Added submodule path detection for `base model`, `lm_head`, and `embeddings` to adapt different base model naming structure;
  - Refactored data loading/preprocessing to support VLM;
- Attn backend improvement:
  - Added option of `sdpa` in case `flex_attn` doesn't work. 
  - Added a unified TTT mask function that produce either `BlockMask` for flex_attn or tensor masks for regular attn.
 - Logging improvements:
   - Added estimated AR validation during training. This is available for both online and offline. 
   - Plot estimated AR and training acc to wandb for better training visualization;
  - Fix: PTQ on speculative decoding model.

## Usage

For VLM as base model, pass in extra arguments `--vlm_processor  <hf_model_path>  --vlm_img_dir <path to images>` in original launching commands. Other usage unchanged.
E.g. 
```bash
./launch_train.sh --model $MODEL \
            --output_dir $OUTPUT_DIR \
            --data $DATA \
            --num_gpu 1 \
            --num_epochs 2 \
            --train_bs 2 \
            --lr 3e-5 \
            --eagle_config eagle_config.json \
            --training_seq_len 4096 \
            --vlm_processor $MODEL \
            --vlm_img_dir  <path to images>
```
## Testing

Tested short training with HF Online training on following models:
- `llama-3.2-1b` -  data: daring-anteater
- The new nano (Hyrbid LLM)  - data: daring-anteater
- The nano-VL - data: Llama-Nemotron-VLM-Dataset-v1/ocr_1
 
See loss decreasing and AR > 1.

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No

### CODE REVIEW

File: modelopt/torch/speculative/plugins/transformers.py [hunk 1]
```diff
@@ -426,34 +472,26 @@ def modify(
...
-            # For offline training, the base model has no layers.
-            # Read the device from the lm_head instead.
-            device = self.lm_head.weight.device
-        elif hasattr(self.model.layers[-1].self_attn, "o_proj"):
-            device = self.model.layers[-1].self_attn.o_proj.weight.device
-        elif hasattr(self.model.layers[-1].self_attn, "q_proj"):
-            device = self.model.layers[-1].self_attn.q_proj.weight.device
-        elif hasattr(self.model.layers[-1].self_attn, "qkv_proj"):
-            device = self.model.layers[-1].self_attn.qkv_proj.weight.device
-        self.eagle_module.to(self.dtype).to(device)
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/transformers.py:439]: TODO: confirm this device detection with @yeyu-nvidia

File: examples/speculative_decoding/eagle_utils.py [hunk 2]
```diff
@@ -107,6 +111,67 @@ def get_role_content(item):
...
+            add_generation_prompt=False,
+        )
+
+        img_filename = os.path.join(img_dir, examples[i]["image"])
+        img = Image.open(img_filename)
+        output = processor(images=img, text=conversation, return_tensors="pt")
+        input_ids = output.input_ids[0]
+        attention_mask = output.attention_mask[0]
+        loss_mask = torch.ones_like(input_ids)
+        labels = torch.full_like(input_ids, IGNORE_TOKEN_ID)
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/eagle_utils.py:161]: So all labels are IGNORE_TOKEN_ID?

File: examples/speculative_decoding/eagle_utils.py
Comment (yeyu-nvidia) [inline examples/speculative_decoding/eagle_utils.py:247]: Does this support VLM data?

File: examples/speculative_decoding/eagle_utils.py [hunk 3]
```diff
@@ -397,24 +508,68 @@ def __call__(self, features: list[dict[str, Any]]) -> dict[str, Any]:
...
+class EagleTrainingPlot(TrainerCallback):
+    """Callback that plot training acc and AR during training."""
+
     def __init__(self, ar_validate_steps: int = 1000):
         self.ar_validate_steps = ar_validate_steps
-        if wandb:
+        if wandb and is_master():
             wandb.init()
 
+    def on_log(self, args, state, control, **kwargs):
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/eagle_utils.py:534]: Can you explain how you estimate AR? I'm not sure it's a good idea to expose "estimated AR" as it may mislead users.

File: examples/speculative_decoding/main.py [hunk 4]
```diff
@@ -75,6 +74,8 @@ class DataArguments:
         default="draft_vocab_cache",
         metadata={"help": "Path to the d2t cache directory."},
     )
+    vlm_img_dir: str = field(default=None, metadata={"help": "Path to the VLM image directory."})
+    vlm_processor: str = field(default=None, metadata={"help": "Path to the VLM processor."})
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/main.py:78]: what is VLM processor?

File: modelopt/torch/speculative/plugins/transformers.py [hunk 5]
```diff
@@ -426,34 +467,26 @@ def modify(
...
-
-        # Make sure self.model.embed_tokens and self.lm_head are frozen
-        for param in self.model.embed_tokens.parameters():
+        # find base model, lm head, and embeddings paths
+        self._find_base_model_parts()
+        self.eagle_module.to(self._base_model.dtype).to(self._base_model_lm_head.weight.device)
```
Comment (yeyu-nvidia) [inline modelopt/torch/speculative/plugins/transformers.py]: Need to check if ptq/inference fails. We want to make sure eagle_module.device is the same as last base model decoder layer, but this is not necessarily the same as lm_head.device.

File: modelopt/torch/speculative/plugins/transformers.py [hunk 6]
```diff
@@ -577,11 +609,29 @@ def msk(b, h, q_idx, kv_idx):
...
                     | ((kv_idx == q_idx + 3 * seq_length) & (kv_idx >= seq_length * 3))
                 )
-
-            return create_block_mask(msk, B=None, H=None, Q_LEN=seq_length, KV_LEN=seq_length * 4)
         else:
             raise ValueError(f"EAGLE TTT step {ttt_step} is not supported")
 
+        dtypemin = torch.finfo(self._base_llm_config.dtype).min
+        q_len = seq_length
+        kv_len = seq_length * (2 + ttt_step)
```
Comment (yeyu-nvidia) [inline modelopt/torch/speculative/plugins/transformers.py:633]: Why 2 + ttt_step?

File: examples/speculative_decoding/eagle_utils.py (hunk 2 above)
Comment (h-guo18) [reply examples/speculative_decoding/eagle_utils.py:161]: Yes. This is aligned with previous behavior:
https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/9c791d9a94c4d92d0a516ead6bd9aeb04e217613/examples/speculative_decoding/eagle_utils.py#L87

File: examples/speculative_decoding/eagle_utils.py
Comment (h-guo18) [reply examples/speculative_decoding/eagle_utils.py:247]: Not yet. Will update in following PR.

File: examples/speculative_decoding/eagle_utils.py (hunk 3 above)
Comment (h-guo18) [reply examples/speculative_decoding/eagle_utils.py:534]: This is calculated by 1 + step_1_acc +  step_1_acc*step_2_acc + step_1_acc*step_2_acc*step_3_acc.

File: examples/speculative_decoding/main.py (hunk 4 above)
Comment (h-guo18) [reply examples/speculative_decoding/main.py:78]: It's a processor that nano-vl use to pre-process the image and the text into tokens. It is defined here:
https://huggingface.co/nvidia/NVIDIA-Nemotron-Nano-12B-v2-VL-BF16/blob/main/processing.py#L43

File: modelopt/torch/speculative/plugins/transformers.py (hunk 6 above)
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/transformers.py:633]: At 0th ttt step, we have kv_len = 2*seq_len

File: examples/speculative_decoding/eagle_utils.py (hunk 2 above)
Comment (yeyu-nvidia) [reply examples/speculative_decoding/eagle_utils.py:161]: but previously we will update labels here https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/9c791d9a94c4d92d0a516ead6bd9aeb04e217613/examples/speculative_decoding/eagle_utils.py#L99

File: examples/speculative_decoding/eagle_utils.py (hunk 3 above)
Comment (yeyu-nvidia) [reply examples/speculative_decoding/eagle_utils.py:534]: I think we have discussed this and agree on not to use estimated AR. We can either just use acc or run real AR validation.

File: modelopt/torch/speculative/plugins/transformers.py (hunk 5 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/transformers.py]: Can you confirm on this?

File: modelopt/torch/speculative/plugins/transformers.py (hunk 6 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/transformers.py:633]: why is that?

File: examples/speculative_decoding/eagle_utils.py (hunk 2 above)
Comment (h-guo18) [reply examples/speculative_decoding/eagle_utils.py:161]: I see. Do you think we will use labels in the future? Otherwise we can get rid of `loss_mask` and `labels` and simplify data loading.

Comment (h-guo18) [reply examples/speculative_decoding/eagle_utils.py:161]: Labels should only be useful when we want to tune the base model. it's currently not used in our code.

Comment (yeyu-nvidia) [reply examples/speculative_decoding/eagle_utils.py:161]: We will need labels (even if it's dummy) as HF trainer needs this for training. We definitely need loss_mask as we need to exclude padded tokens. I would say we keep them.

File: modelopt/torch/speculative/plugins/transformers.py (hunk 5 above)
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/transformers.py]: I see the idea. I think it make sense to put eagle on last layer's device.

File: examples/speculative_decoding/eagle_utils.py (hunk 3 above)
Comment (h-guo18) [reply examples/speculative_decoding/eagle_utils.py:534]: Disabled estimate AR by default. I kept it in code since it's useful for experiments.

**Review by yeyu-nvidia — APPROVED** [review]
